go/fn/errors.go

```diff
@@ -71,3 +41,20 @@ type ErrInternalAnnotation struct {
 func (e *ErrInternalAnnotation) Error() string {
 	return e.Message
 }
+func NewErrUnmatchedField(obj SubObject, fields []string, dt any) *ErrUnmatchedField {
```

Comment:
nit: Can we add a doc comment to this, and the below `ErrUnmatchedField` type?

Reply:
Done. also changed the arg `dt` to `expectedFieldType`